src/adapters/net.cpp

```diff
@@ -28,6 +30,38 @@ namespace net {
     return file_util::exists("/sys/class/net/" + ifname + "/wireless");
   }
 
+  std::string find_wireless_interface() {
+    DIR* dir = opendir("/sys/class/net");
+    while (true) {
+      struct dirent* entry = readdir(dir);
+      if (entry == nullptr) {
+        break;
+      }
+      if (is_wireless_interface(entry->d_name)) {
+        closedir(dir);
+        return entry->d_name;
+      }
+    }
+    closedir(dir);
+    return "";
+  }
+
+  std::string find_wired_interface() {
+    DIR* dir = opendir("/sys/class/net");
+    while (true) {
+      struct dirent* entry = readdir(dir);
+      if (entry == nullptr) {
+        break;
+      }
+      if (entry->d_name[0] != '.' && !is_wireless_interface(entry->d_name)) {
```

Comment:
This also picks up the `lo` interface which is almost certainly not what the user wants. It will also pick up things like vpn interfaces and so on (anything that isn't wireless).

I think if "wired" is specified it should only pick up ethernet interfaces.

I don't think there is a way to perfectly capture the ethernet interface on all systems (at least not without major effort). But that's okay, this kind of function only has to be good enough to detect some wired interface for most people.

i3status has a very involved approach to that in [`first_network_device.c`](https://github.com/i3/i3status/blob/master/src/first_network_device.c), we don't have to do everything they do, but I think one thing to check for is whether the interface is virtual or not and only accept it as a wired or wireless interface if it is not virtual. An interface is virtual, if its entry in `/sys/class/net` is symlinked to a directory in `/sys/devices/virtual/`.

Reply:
I had been looking at i3status as well, and thinking maybe I could get away with just some simple fs checks, but combined with the readdir race condition, I think I will take the getifaddrs approach.

Reply:
That's a good idea too :)